### DISCUSSION TITLE
Support bundle verification with multiple public keys

### DISCUSSION THREAD

**Discussion by baka3k** [callstack/repack#823] [Ideas]
The situation we're facing is as follows:
We have multiple independent MiniApp development teams, and we want each team to use their own unique private key.

![SuperApp_Diagram-Signature](https://github.com/user-attachments/assets/86863fc6-7ff1-43d4-babe-18e0136cbbd9)

This means that when verifying a bundle, we need to pass in the public key of the respective team (which is fetched from the server).

![SuperApp_Diagram-Verification](https://github.com/user-attachments/assets/5c141569-1361-42b6-bf0a-7ae57e002904)

However, when examining the repack source code, we found that the 'verify' function is currently only allowing for a single public key that is hardcoded into the source code, as shown below."
```
fun verifyBundle(context: Context, token: String?, fileContent: ByteArray?) {
            if (token == null) {
                throw Exception("The bundle verification failed because no token for the bundle was found.")
            }

            val stringPublicKey = getPublicKeyFromStringsIfExist(context)
                    ?: throw Exception("The bundle verification failed because PublicKey was not found in the bundle. Make sure you've added the PublicKey to the res/values/strings.xml under RepackPublicKey key.")

            val publicKey = parsePublicKey(stringPublicKey)
                    ?: throw Exception("The bundle verification failed because the PublicKey is invalid.")

            val claims: Map<String, Any?> = verifyAndDecodeToken(token, publicKey)

            val contentHash = claims["hash"] as String?
                    ?: throw Exception("The bundle verification failed because the token is invalid.")

            val fileHash = computeHash(fileContent)

            if (contentHash != fileHash) {
                throw Exception("The bundle verification failed because the bundle hash is invalid.")
            }
        }

```
```
 private fun getPublicKeyFromStringsIfExist(
                context: Context
        ): String? {
            val packageName: String = context.packageName
            val resId: Int =
                    context.resources.getIdentifier("RepackPublicKey", "string", packageName)
            if (resId != 0) {
                return context.getString(resId).ifEmpty {
                    null
                }
            }
            return null
        }
```
It would be much better if the verify bundle function could accept an additional public key as a parameter, instead of hardcoding it as it is now, such as:
```
fun verifyBundle(context: Context, 
                         token: String?, 
                         fileContent: ByteArray?,
                         publicKey: ByteArray? // add this parameter
) {
```
With this approach, we can have multiple mini-app development teams, each with their own private key for signing their mini-apps. We can authenticate development teams using their keys, allowing them to join the ecosystem or rejecting teams without relying on the main app's private key.
Furthermore, if the main app's private key is lost or compromised, mini-apps created by various development teams can still operate. We have the flexibility to re-sign them whenever needed,(similar Google's Play App Signing approach)

Would you please consider this proposal

Thanks & BestRegard,

**Comment by jbroma** [comment]
hey @baka3k can you please recreate this as an issue with a template for a feature request? Thanks!

**Reply by baka3k** [reply]
Let's continue this discussion in ticket https://github.com/callstack/repack/issues/826
